src/wasm-stack.h

```diff
@@ -197,10 +197,21 @@ template<typename SubType> void BinaryenIRWriter<SubType>::write() {
   emitFunctionEnd();
 }
 
-// emits a node, but if it is a block with no name, emit a list of its contents
+// Emits a node in a position that can contain a list of contents, like an if
+// arm. This will emit the node, but if it is a block with no name, just emit
+// its contents. This is ok to do because a list of contents is ok in the wasm
+// binary format in such positions anyhow. When we read such code in Binaryen
+// we will end up creating a block for it (note that while doing so we create a
+// block without a name, since nothing branches to it, which makes it easy to
+// handle in optimization passes and when writing the binary out again).
 template<typename SubType>
 void BinaryenIRWriter<SubType>::visitPossibleBlockContents(Expression* curr) {
   auto* block = curr->dynCast<Block>();
+  // Even if the block has a name, check if the name is necessary (if it has no
+  // uses, it is equivalent to not having one). Scanning the children of the
+  // block means that this takes quadratic time, but it will be N^2 for a fairly
+  // small N since the number of nested non-block control flow structures tends
+  // to be very reasonable.
```

Comment:
We could still make this cheaper and also improve the later case by running a pass to remove unused names in linear time, then assuming we've already taken care of it here. We already have a pass that does this, don't we?

Reply:
Yes, RemoveUnusedNames is run when optimizing. So this only matters for the unoptimized case. But it's still nice to handle that since that includes a simple roundtrip with wasm-opt.

Reply:
> since that includes a simple roundtrip with wasm-opt

I don't understand what you mean here. I see that this quadratic checking will only produce useful results if we haven't already optimized, but it would still be cheaper and produce better results to run the `RemoveUnusedNames` pass unconditionally here no matter what other optimizations we are doing.

Reply:
Yeah, running the pass might be fine. We'd need to do some more work to be careful to avoid noticeable side effects, as right now binary writing does not modify the Module data, but running that pass would. But I think it could be done with some work.

Reply:
Looking at RemoveUnusedBrs, it looks like it does more optimizations than we would actually want for this use case. What Module data are you thinking about, though? That pass is function parallel, so I don't think it should change any Module-level state.

Reply:
I meant any changes to the Module or its contents. Right now, writing a module does not modify the input in any way, but this would make changes to blocks etc. Still, it might be ok...

Reply:
Oh, I see. Yeah, I think that's ok since there shouldn't really be anything happening after writing. This PR lgtm if you'd rather look at that as a follow-up.

Reply:
Yeah, let's leave that as a separate topic. It's the existing behavior before this PR and I'd rather not change it here so each PR is focused.